fix: cp-7.50.0 Earn lending blocking UI tweaks

nickewansmith opened the pull request:
## **Description**
1. What is the reason for the change?
- There are some UIX issues noticed during lending flow that should be addressed before the release

2. What is the improvement/solution?
- Update styles to accommodate smaller phones
- Add polling to earn token list to pull in cross network tokens
- Ensure lending learn more modal always uses the underlying token context
- Ensure that the empty state CTA switches to the lending token network
- Improve lending confirmation view handling
- Remove position dependency to further increase response speed and reliability

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-mobile/issues/16834
Fixes: https://consensyssoftware.atlassian.net/browse/TAT-137

## **Manual testing steps**

- General testing of the earn lending feature
- Deposit/Withdraw/Approve
- Popular Networks / Single Network actions
- On small phones the feature should not have hidden buttons or cut off views
- Learn more modal should list earn token symbol only for all popups (withdraw and receipt token asset details earn section)

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

https://www.loom.com/share/343b169bc4e24d6997058c1f25c1410a?sid=c884a764-2b1c-4fdf-9180-322ec7a7d994

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

nickewansmith committed:
fix: Update styles to accommodate smaller phones

- Addresses earn token list overlapping the currency switcher
- Address various input forms or confirmation modals with buttons that were off the screen
- Make tooltip areas larger to avoid missing touch events
- Fix how it works title too small in learn more modal
- Fix token logo rendering for native tokens
- Eliminate unnec space and correct padding in token list, ctas
- Adjust token names to have ellipses if too long and work for smaller phones as well
- Add scroll view where needed to allow smaller phones to see the entire lending input form

nickewansmith committed:
fix: Add polling to earn token list to pull in cross network tokens

- When not on the popular networks filter view, the tokens from other networks are not detected on boot. We add a polling mechanism that can start this process once the EarnTokenList is viewed so we can see tokens across networks once loading that view, stopping the polling unmount
- Use “Earn up to 4.5%” only when user has no tokens eligible to earn, otherwise use the initially designed CTA
- Order withdrawals token list by highest balance
- Update titles for Deposit and Withdraw actions

nickewansmith committed:
fix: Ensure lending learn more modal always uses the underlying token context

- Ensures that the learn more modal displays the underlying token symbol regardless of the asset passed

nickewansmith committed:
fix: Ensure that the empty state CTA switches to the lending token network

nickewansmith committed:
fix: Improve lending confirmation view handling

- Graceful error handling for temp lending utils
- Ensure navigation to Transaction View after tx submitted as certain networks have a race condition that redirects to Home instead
- Stabilize the maxWithdrawalAmount logic so that it does not return a value when a user does not have borrow positions

nickewansmith committed:
fix: Remove position dependency to further increase response speed and reliability

nickewansmith committed:
fix: Add missing snapshot

nickewansmith commented on app/components/UI/Earn/components/EmptyStateCta/index.tsx:
```diff
@@ -58,7 +59,21 @@ const EarnEmptyStateCta = ({ token }: EarnEmptyStateCta) => {
   ).toFixed(0);
   const apr = earnToken?.experience?.apr;
-  const navigateToLendInputScreen = () => {
+  const navigateToLendInputScreen = async () => {
```
This is on line 93
```
navigate('StakeScreens', {
      screen: Routes.STAKING.STAKE,
      params: { token },
    });
    ```

nickewansmith commented on app/components/UI/Earn/utils/token/index.ts:
```diff
@@ -76,6 +76,11 @@ export const sortByHighestApr = (tokensToSort: EarnTokenDetails[]) =>
 
+export const sortByHighestBalance = (tokensToSort: EarnTokenDetails[]) =>
+  [...tokensToSort].sort((a, b) =>
+    b.balanceFiatNumber > a.balanceFiatNumber ? 1 : -1,
```
This is fine in this case

Matt561 commented on app/components/UI/Earn/hooks/useEarnNetworkPolling.ts:
```diff
@@ -0,0 +1,123 @@
+        const chainDetectedTokens =
+          allDetectedTokens[chainId]?.[selectedAccount.address];
+        if (
+          chainDetectedTokens &&
```
Should we skip if the token list already has the token imported?

nickewansmith replied:
These tokens can be detected at any point in time.

Matt561 commented on app/components/UI/Earn/hooks/useEarnNetworkPolling.ts:
```diff
@@ -0,0 +1,123 @@
+ *
+ * This ensures that lending tokens from all supported networks are available
+ */
+export const useEarnNetworkPolling = () => {
```
I don't understand the problem you're solving as much so this may be naive _but_ would this make more sense to call once on app boot instead of every time a component using this hook is rendered? Is there a reason to call this more than just on app boot? Maybe the `tokensDetected` is updated every time a new token with a non-zero balance is found?

Apologies if this is a dumb question

nickewansmith replied:
We could call this on app boot, but this would add extra load to the app. When the app starts for a new user it starts on mainnet ETH. I don't want to introduce polling for other chains on app boot before they click on the popular networks filter. Instead we only use it when the earn token list is being viewed. The polling controller is handling de-duplication of polling input calls behind the scene.

> Maybe the tokensDetected is updated every time a new token with a non-zero balance is found?
This already happens due to global polling, but it won't happen for tokens on other networks when on the single network view as how the global pollers are currently programmed. This hook ensures that that is never a problem.

Matt561 commented on app/components/UI/Earn/utils/tempLending.ts:
```diff
@@ -376,14 +376,15 @@ export const getAaveV3MaxSafeWithdrawal = async (
       // assetPrice is in USD with 8 decimals
       // We want result in token decimals
-      const maxTokenAmount = totalCollateralBase
-        .mul(parseUnits('1', tokenDecimals)) // Scale up to token decimals
-        .div(assetPrice); // Divide by price (removes the 8 decimals from price)
+      // const maxTokenAmount = totalCollateralBase
+      //   .mul(parseUnits('1', tokenDecimals)) // Scale up to token decimals
+      //   .div(assetPrice); // Divide by price (removes the 8 decimals from price)
 
-      return maxTokenAmount.toString();
+      // return maxTokenAmount.toString();
```
Let's delete this commented out code If we don't need to calculate the total collateral since zero debt always means the health factor won't be impacted afaik.

Matt561 commented on app/components/UI/Earn/utils/tempLending.ts:
```diff
@@ -657,6 +663,8 @@ export const getAaveV3MaxRiskAwareWithdrawalAmount = async (
         getAaveV3MaxSafeWithdrawal(userData, receiptToken as EarnTokenDetails),
       ]);
 
+    if (!maxHealthFactorWithdrawalInTokens) return undefined;
```
We likely don't want this since we still want to check poolLiquidity. If the `maxHealthFactorWithdrawalInTokens` is undefined. We can filter out undefined values before calling `BigNumber.min`

```ts
const values = [value1, value2, value3].filter(v => v !== undefined);
const min = BigNumber.min(...values);
```

nickewansmith replied:
I have added back the pool liquidity check on withdraw

Matt561 commented on app/components/UI/Earn/components/EarnTokenList/index.tsx:
```diff
@@ -87,6 +88,10 @@ export interface EarnTokenListProps {
 const EarnTokenList = () => {
+  // Start polling lending networks when this component mounts and stops when it unmounts
+  // This is currently the main component that needs data cross chain on boot
+  useEarnNetworkPolling();
```
I think it would be valuable to explain why this hook is necessary in the hook's file for future us.

Matt561 commented on app/components/UI/Earn/components/EarnTokenSelector/EarnTokenSelector.styles.ts:
```diff
@@ -16,11 +16,13 @@ export default (params: { theme: Theme }) => {
       flexDirection: 'row',
-      maxWidth: 210,
       alignItems: 'center',
+      maxWidth: '95%',
```
These percentage values worry me a bit. I've had headaches in the past getting consistent UIs across screen sizes with this approach.

nickewansmith replied:
Understood, they would worry me more as well if I wasn't testing with 3 distinct phone sizes. These percentages work better than what existed before. Anywhere I add these style changes are to accommodate android/ios small/medium/large. I'm sure it is not perfect, but the previous ones were not working well with longer text on either side. I am no css guru, but this does test well on the devices I used and is a minimal surface area.

nickewansmith commented:
https://cursor.com/agents?selectedBcId=bc-1be99df1-f75e-4635-95fe-b54d72a51b52

nickewansmith commented on app/components/UI/Earn/Views/EarnInputView/EarnInputView.tsx:
```diff
@@ -774,40 +775,48 @@ const EarnInputView = () => {
-          />
-        )}
-      </View>
+      <ScrollView
```
The scroll view is not on the keypad, only the top portion which has a lot of possible content. Instead of getting into device conditions, I just did this temporarily for the phones that are small so everything is still functional. On larger phones the view up top does not scroll.

nickewansmith commented on app/components/UI/Earn/Views/EarnLendingDepositConfirmationView/components/DepositInfoSection/index.tsx:
```diff
@@ -87,12 +87,16 @@ const DepositInfoSection = ({
                   }
                 </Text>
-                <Text color={TextColor.Alternative}>
+                <Text
+                  color={TextColor.Alternative}
+                  numberOfLines={1}
+                  ellipsizeMode="head"
+                >
```
On smaller phones and even on larger ones that row does not have that much space. If estimated annual rewards were $900.56 983.45677 USDC. We must do something to stay on the line. These would be rarer cases. I thought it safer to have these ellipses vs making the text split to two rows without design decision

nickewansmith replied:
Example output $900.68 ...65 USDC

nickewansmith commented on app/components/UI/Earn/Views/EarnLendingDepositConfirmationView/index.tsx:
```diff
@@ -379,7 +379,10 @@ const EarnLendingDepositConfirmationView = () => {
           emitDepositTxMetaMetric(MetaMetricsEvents.EARN_TRANSACTION_SUBMITTED);
-          navigation.navigate(Routes.TRANSACTIONS_VIEW);
+          // There is variance in when navigation can be called across chains
+          setTimeout(() => {
+            navigation.navigate(Routes.TRANSACTIONS_VIEW);
+          }, 0);
```
I did try this, but it didn't work. This race condition seems to happen on faster networks and would require more investigation to get to the bottom of why that is happening.

nickewansmith commented on app/components/UI/Earn/Views/EarnWithdrawInputView/EarnWithdrawInputView.tsx:
```diff
@@ -151,6 +152,7 @@ const EarnWithdrawInputView = () => {
       !receiptToken?.chainId
     )
       return;
+    setMaxRiskAwareWithdrawalAmount(undefined);
```
Yes this fixes that bug by unsetting the max withdrawal amount in the parent caller when the user has not borrowed anything. Then if they have borrowed we do the other checks and return a value.

nickewansmith commented on app/components/UI/Earn/components/EarnTokenList/__snapshots__/EarnTokenList.test.tsx.snap:
```diff
@@ -175,7 +175,7 @@ exports[`EarnTokenList render matches snapshot 1`] = `
               }
             >
-              Select a token
+              Select a token to deposit
```
Yes, it renders the withdrawal text when on the withdrawal list

nickewansmith committed:
chore(feedback): Add documentation for useEarnNetworkPolling hook

- Adds documentation for the new useEarnNetworkPolling hook
- Updates tempLending methods to also consider pool liquidity if possible and updates related unit tests
